src/loaders/stm_load.c

```diff
@@ -269,10 +337,28 @@ static int stm_load(struct module_data *m, HIO_HANDLE * f, const int start)
 	/* Read samples */
 	D_(D_INFO "Stored samples: %d", mod->smp);
 
+    	if (m->filename && (x = strrchr(m->filename, '/')))
+	    strncpy(pathname, m->filename, x - m->filename);
+
 	for (i = 0; i < mod->ins; i++) {
-		if (mod->xxs[i].len > 1) {
-			if (libxmp_load_sample(m, f, 0, &mod->xxs[i], NULL) < 0)
-				return -1;
+		if (sfh.ins[i].volume && sfh.ins[i].length) {
+			if (sfh.type == STM_TYPE_SONG) {
+			    HIO_HANDLE *s;
+			    char sn[256];
+			    snprintf(sn, XMP_NAME_SIZE, "%s%s", pathname, mod->xxi[i].name);
```

Comment:
This should probably guard against path traversal issues by not allowing characters such as / and \ in the filename. Maybe only accept alphanumerics in filenames to completely minimize the risk posed by malicious files.

Reply:
I have copied this piece of code from mod_load.c

Reply:
Okay, then Claudio should review this probably in all similar places, not just here.